src/plugins/intel_cpu/src/nodes/scaled_attn.cpp

```diff
@@ -806,6 +824,288 @@ bool ScaledDotProductAttention::isSupportedOperation(const std::shared_ptr<const
+            no_reorder = false;
+            break;
+        }
+    }
+
+    // reorder
+    if (!no_reorder) {
+        m_tmp_reorder.resize<int32_t>({B, L0});
+        for (size_t i = 0; i < B; i++) {
+            std::memcpy(&m_tmp_reorder.at<int32_t>({i}),
+                        &beam_table_k.at<int32_t>({i}),
+                        sizeof(int32_t) * L0);
+        }
+        auto* table = beam_idx.data<int32_t>();
+        // beam table is same for both k,v state
+        for (size_t i = 0; i < B; i++) {
+            std::memcpy(&beam_table_k.at<int32_t>({i}),
+                        &m_tmp_reorder.at<int32_t>({static_cast<size_t>(table[i])}),
+                        sizeof(int32_t) * L0);
+            std::memcpy(&beam_table_v.at<int32_t>({i}),
+                        &m_tmp_reorder.at<int32_t>({static_cast<size_t>(table[i])}),
+                        sizeof(int32_t) * L0);
+        }
+    }
+    // second token itself
+    for (size_t i = 0; i < B; i++) {
+        for (size_t j = 0; j < L1; j++) {
+            beam_table_k.at<int32_t>({i, L0 + j}) = i;
+            beam_table_v.at<int32_t>({i, L0 + j}) = i;
+        }
+    }
+}
+
+void ScaledDotProductAttention::updatePastkv(const MemoryPtr& mem_cur_k, const MemoryPtr& mem_cur_v) {
+    std::vector<size_t> order = {0, 1, 2, 3};
+    if (!m_config.config.permute_axes.empty()) {
+        order = m_config.config.permute_axes;
+    }
+    PlainTensor cur_k, past_k;
+    PlainTensor cur_v, past_v;
+    cur_k.reset(mem_cur_k);
+    cur_v.reset(mem_cur_v);
+    cur_k = cur_k.permute(order);
+    cur_v = cur_v.permute(order);
+    auto B = cur_k.size(0);
+    auto H = cur_k.size(1);
+    auto L1 = cur_k.size(2);
+    auto S = cur_k.size(3);
+    size_t L0 = 0;
+    auto& reverse_order = m_config.reverse_order;
+    auto reverse = [&reverse_order] (const std::vector<size_t>& cur) {
+        std::vector<size_t> result(cur.size());
+        for (size_t i = 0; i < cur.size(); i++) {
+            result[i] = cur[reverse_order[i]];
+        }
+        return result;
+    };
+    auto internal_mem_k = m_k_state->internal_state_mem();
+    auto internal_mem_v = m_v_state->internal_state_mem();
+
+    auto is_reset = m_k_state->is_reset_state();
+    if (is_reset) {
+        auto inputNumber = getOriginalInputsNumber();
+        auto&& init_graph_v_dims = getParentEdgeAt(inputNumber - 1)->getMemory().getStaticDims();
+        L0 = init_graph_v_dims.at(order[2]);
+        auto B_init = init_graph_v_dims.at(order[0]);
+        OPENVINO_ASSERT(B == B_init, "pastkv batch: ", B, " is not equal to batch of init subgraph: ", B_init);
+    } else if (internal_mem_k) {
+        auto block_desc = internal_mem_k->getDescWithType<BlockedMemoryDesc>();
+        auto&& dims = block_desc->getShape().getStaticDims();
+        auto B_state = dims[order[0]];
+        L0 = dims[order[2]];
+        OPENVINO_ASSERT(B == B_state, "pastkv batch: ", B, " is not equal to batch of state: ", B_state);
+    }
```

Comment:
MemoryInputSDPA  node can always give correct shape for `past_key/value`, whether it's just reset or not, so why not simply retrieve L0 from input edge?

Reply:
Updated.